### PR TITLE
Integrate with new lora_plan module

### PR DESCRIPTION
**Pull request by mikev**

### CODE REVIEW

File: src/device/router_device_worker.erl [hunk 1]
```diff
@@ -962,7 +963,7 @@ handle_info(
     Device1 = router_device:metadata(Metadata, Device),
     DownlinkPacket = blockchain_helium_packet_v1:new_downlink(
         craft_join_reply(Device1, JoinAcceptArgs),
-        lorawan_mac_region:downlink_signal_strength(Region, TxFreq),
+        lora_plan:downlink_eirp(lora_plan:region_to_plan(Region), TxFreq),
```
Comment (michaeldjeffrey) [inline src/device/router_device_worker.erl:966]: This is one where I'm wondering if we should keep the `downlink_signal_strength` name. Or alias it at the bottom of this file somewhere. 

Going to the spec for making a new downlink gets you
```erlang
-spec new_downlink(Payload :: binary(), SignalStrength :: integer(), Timestamp :: non_neg_integer(),
                   Frequency :: float(), DataRate :: string(), Rx2 :: window() | undefined) -> packet().
```

You kind of have to just _know_ what EIRP is already if you come across this.

File: src/device/router_device_worker.erl [hunk 2]
```diff
@@ -1888,9 +1892,12 @@ channel_correction_and_fopts(Packet, Region, Device, Frame, Count, ADRAdjustment
                 %% `lorwan_mac_region'.  But `set_channels' wants data
                 %% rate in the form of "SFdd?BWddd?" so that's what
                 %% we'll give it.
-                NewDr = lorawan_mac_region:dr_to_datar(Region, NewDataRateIdx),
+                Plan = lora_plan:region_to_plan(Region),
+                NewDr = lora_plan:atom_to_datarate(
+                    (lora_plan:index_to_datarate(Plan, NewDataRateIdx))
+                ),
```
Comment (michaeldjeffrey) [inline src/device/router_device_worker.erl:1898]: This part seems a bit weird. Both are `_to_datarate` functions, but return different forms of datarate?
I'm not sure the best way to disambiguate "datarate" here.

**Review by dpezely — DISMISSED** [review]

File: src/device/router_device_worker.erl [hunk 3]
```diff
@@ -1902,9 +1909,9 @@ channel_correction_and_fopts(Packet, Region, Device, Frame, Count, ADRAdjustment
         case lists:member(link_check_req, Frame#frame.fopts) of
             true ->
                 SNR = blockchain_helium_packet_v1:snr(Packet),
-                MaxUplinkSNR = lorawan_mac_region:max_uplink_snr(
-                    list_to_binary(blockchain_helium_packet_v1:datarate(Packet))
-                ),
+                DataRateStr = list_to_binary(blockchain_helium_packet_v1:datarate(Packet)),
+                DataRateAtom = lora_plan:datarate_to_atom(DataRateStr),
```
Comment (michaeldjeffrey) [inline src/device/router_device_worker.erl:1913]: Is there a way we could prevent the following from bypassing the lib and going directly do `max_uplink_snr/1`?

```erlang
erlang:list_to_atom(blockchain_helium_packet_v1:datarate(Packet))
```

Going to the spec
```erlang
-spec max_uplink_snr(atom()) -> float().
```
all we see is the need of an atom.

File: src/lora/lorawan_adr.erl [hunk 4]
```diff
@@ -291,19 +291,13 @@
...
-    Datarates = lists:filtermap(FilterMapFn, lorawan_mac_region:datars(Region)),
-    %% min-max refer to #device.min_datarate docs
-    [{MinDataRate, {MaxSpreading, _}} | _] = Datarates,
-    {MaxDataRate, {MinSpreading, _}} = lists:last(Datarates),
-    TxPowers = lorawan_mac_region:uplink_power_table(Region),
+    MinSF = min_adr_sf(Region),
+    MaxSF = max_adr_sf(Region),
+    _MinDataRate = min_adr_dr(Region),
+    MaxDataRate = max_adr_dr(Region),
+    Datarates = [{1, {7, 125}}],
```
Comment (michaeldjeffrey) [inline src/lora/lorawan_adr.erl:298]: If these values work for all the ADR plans we support right now, can we move it into a macro?

File: src/lora/lorawan_adr.erl [hunk 5]
```diff
@@ -291,19 +291,13 @@
...
-            false
-    end,
-    Datarates = lists:filtermap(FilterMapFn, lorawan_mac_region:datars(Region)),
-    %% min-max refer to #device.min_datarate docs
-    [{MinDataRate, {MaxSpreading, _}} | _] = Datarates,
-    {MaxDataRate, {MinSpreading, _}} = lists:last(Datarates),
-    TxPowers = lorawan_mac_region:uplink_power_table(Region),
+    MinSF = min_adr_sf(Region),
+    MaxSF = max_adr_sf(Region),
+    _MinDataRate = min_adr_dr(Region),
```
[the remark this answers is not in the document]
Comment (mikev) [reply src/lora/lorawan_adr.erl]: Changed function to adr_datarate_range

File: src/lora/lorawan_adr.erl [hunk 6]
```diff
@@ -314,16 +308,36 @@ new(Region) ->
         accepted_adjustments = [],
         datarates = Datarates,
         txpowers = TxPowers,
-        min_datarate = MinDataRate,
-        max_spreading = MaxSpreading,
+        min_datarate = 1,
```
[the remark this answers is not in the document]
Comment (mikev) [reply src/lora/lorawan_adr.erl]: No, the type was set to pos_integer().  Fixed.

File: src/device/router_device_worker.erl (hunk 1 above)
Comment (mikev) [reply src/device/router_device_worker.erl:966]: I've added a new API named max_tx_power.  We'll replace downlink_eirp with max_tx_power.

Comment (mikev) [reply src/device/router_device_worker.erl]: Replaced with max_tx_power.

File: test/router_lorawan_SUITE.erl [hunk 7]
```diff
@@ -263,12 +263,13 @@ lw_join_test(Config) ->
...
 
     %% Waiting for reply resp form router
-    {_NetID, _DevAddr, _DLSettings, _RxDelay, NwkSKey, AppSKey, CFList} = test_utils:wait_for_join_resp(
+    {_NetID, _DevAddr, _DLSettings, _RxDelay, NwkSKey, AppSKey, _CFList} = test_utils:wait_for_join_resp(
         PubKeyBin,
         AppKey,
         DevNonce
     ),
-    ?assertEqual(CFList, lorawan_mac_region:mk_join_accept_cf_list(Region)),
+    % ToDo: Fixme
```
Comment (dpezely) [inline test/router_lorawan_SUITE.erl:271]: Would you elaborate on this TODO item, please?

Comment (mikev) [reply test/router_lorawan_SUITE.erl:271]: I'll need some CT expert help on this one.

Comment (mikev) [reply test/router_lorawan_SUITE.erl]: Added a print statement and the problem went away.  Go figure; problem solved.

Note - there are also eunit tests which verify that join_cf_list is working as expected.

**Review by dpezely — DISMISSED** [review]
Looks good.  There are still a couple of unresolved conversation threads, but I'm fine with marking those as TODO items.

File: src/device/router_device_worker.erl (hunk 2 above)
Comment (mikev) [reply src/device/router_device_worker.erl]: I've changed the API to these three APIs

datarate_to_index(Plan, DataRate)
datarate_to_atom(Plan, DataRate)
datarate_to_binary(Plan, DataRate)

The APIs will handle all three cases.  Definitely makes the code clearer.

**Comment by macpie** [comment]
Seems like there must be a log line somewhere in the lora lib: 
```
datarate_to_index Binary=<<83,70,56,66,87,49,50,53>>
datarate_to_index Atom='SF8BW125'
...
```

Could we try to remove it? Looks like an `io:format/2`

File: src/device/router_device_worker.erl [hunk 8]
```diff
@@ -1830,7 +1834,7 @@ handle_frame_timeout(
 ) -> {boolean(), list()}.
 channel_correction_and_fopts(Packet, Region, Device, Frame, Count, ADRAdjustment) ->
     ChannelsCorrected = were_channels_corrected(Frame, Region),
-    DataRate = blockchain_helium_packet_v1:datarate(Packet),
+    DataRateBinary = list_to_binary(blockchain_helium_packet_v1:datarate(Packet)),
```
Comment (michaeldjeffrey) [inline src/device/router_device_worker.erl:1837]:
```suggestion
    DataRateBinary = erlang:list_to_binary(blockchain_helium_packet_v1:datarate(Packet)),
```

If only to keep with convention of using builtins in this repo.

File: src/device/router_device_worker.erl [hunk 9]
```diff
@@ -2142,8 +2149,9 @@ maybe_track_adr_packet(Device, ADREngine0, FrameCache) ->
         %% power this packet transmitted at, but that value is
         %% unknowable.
         {false, true} ->
-            DataRateStr = blockchain_helium_packet_v1:datarate(Packet),
-            DataRateIdx = lorawan_mac_region:datar_to_dr(Region, DataRateStr),
+            DataRateBinary = list_to_binary(blockchain_helium_packet_v1:datarate(Packet)),
```
Comment (michaeldjeffrey) [inline src/device/router_device_worker.erl:2152]:
```suggestion
            DataRateBinary = erlang:list_to_binary(blockchain_helium_packet_v1:datarate(Packet)),
```

File: src/lora/lorawan_mac_commands.erl [hunk 10]
```diff
@@ -425,7 +427,9 @@ find_status(FOptsIn) ->
 
 send_link_check([{_MAC, RxQ} | _] = Gateways) ->
     #rxq{datr = DataRate, lsnr = SNR} = RxQ,
-    Margin = trunc(SNR - lorawan_mac_region:max_uplink_snr(DataRate)),
+    DataRateAtom = binary_to_atom(DataRate),
```
Comment (michaeldjeffrey) [inline src/lora/lorawan_mac_commands.erl:430]:
```suggestion
    DataRateAtom = erlang:binary_to_atom(DataRate),
```

File: src/lora/lorawan_mac_commands.erl [hunk 11]
```diff
@@ -425,7 +427,9 @@ find_status(FOptsIn) ->
...
+    MaxSNR = lora_plan:max_uplink_snr(DataRateAtom) + 10,
+    Margin = trunc(SNR - MaxSNR),
```
Comment (michaeldjeffrey) [inline src/lora/lorawan_mac_commands.erl:432]:
```suggestion
    Margin = erlang:trunc(SNR - MaxSNR),
```

File: src/lora/lorawan_mac_region.erl [hunk 12]
```diff
@@ -1,1594 +0,0 @@
...
-%% @doc
-%% Copyright (c) 2016-2019 Petr &lt;Gotthard petr.gotthard@@centrum.cz&gt;
-%% All rights reserved.
-%% Distributed under the terms of the MIT License. See the LICENSE file.
-%% @end
-%%%-------------------------------------------------------------------
--module(lorawan_mac_region).
-
-%% Functions that map Region -> Top Level Region
--export([
```
Comment (michaeldjeffrey) [inline src/lora/lorawan_mac_region.erl:11]: Good riddance. You love to see it. ❤️

File: test/router_data_SUITE.erl [hunk 13]
```diff
@@ -454,8 +454,8 @@ data_test_3(Config) ->
             <<"hotspot">> => #{
                 <<"id">> => erlang:list_to_binary(libp2p_crypto:bin_to_b58(PubKeyBin)),
                 <<"name">> => erlang:list_to_binary(HotspotName),
-                %% RSSI act as power here we are and based on UPLINK RSSI < -80 we should power downlink at 27
-                <<"rssi">> => 27,
+                %% RSSI act as power here we are and based on UPLINK RSSI < -80 we should power downlink at 20
+                <<"rssi">> => 20,
```
Comment (michaeldjeffrey) [inline test/router_data_SUITE.erl:458]: This comment is a bit confusing to me when read with the one above.
Most of the RSSI's in the tests are changing to 30.
These two are the outliers, and I don't understand from the comments why they're both changing to 20 under opposite circumstances.

```
Above comment from L315
%% RSSI act as power here we are and based on UPLINK RSSI > -80 we should power downlink at 20

vs. this comment L457
%% RSSI act as power here we are and based on UPLINK RSSI < -80 we should power downlink at 20
```

File: test/router_lorawan_SUITE.erl [hunk 14]
```diff
@@ -268,7 +268,10 @@ lw_join_test(Config) ->
         AppKey,
         DevNonce
     ),
-    ?assertEqual(CFList, lorawan_mac_region:mk_join_accept_cf_list(Region)),
+    ct:print(error, ?STD_IMPORTANCE, "CFList = = ~w", [CFList]),
+    ChMaskList = lora_chmask:join_cf_list(Region),
+    ct:print(error, ?STD_IMPORTANCE, "ChMaskList = = ~w", [ChMaskList]),
```
Comment (michaeldjeffrey) [inline test/router_lorawan_SUITE.erl:273]:
```suggestion
    ChMaskList = lora_chmask:join_cf_list(Region),
```

**Review by michaeldjeffrey — COMMENTED** [review]
I'm excited for this to land.

File: test/router_data_SUITE.erl (hunk 13 above)
Comment (mikev) [reply test/router_data_SUITE.erl:458]: Lets pair code review this tomorrow.

**Review by michaeldjeffrey — APPROVED** [review]

File: test/router_data_SUITE.erl (hunk 13 above)
Comment (michaeldjeffrey) [reply test/router_data_SUITE.erl:458]: We found this to be a misunderstanding of the way gateway regions and device regions are used to compute the downlink power.

**Review by macpie — APPROVED** [review]

**Pull request merged**